### ISSUE TITLE
ip_local_reserved_ports should be set to deconflict with ip_local_port_range 

### ISSUE THREAD

**Issue by ghost** [cloudfoundry/routing-release#206]
## Issue

If you're unlucky, an unrelated process on a Gorouter can consume a local port that the Gorouter needs to be able to listen on, causing the Gorouter to be stuck in a restart loop.

## Affected Versions

Likely all in the past 5 years.

## Context

https://github.com/cloudfoundry/routing-release/commit/e00e0ca5f8ae3cc9f02abc81642b45a6a38f648d adds the ability to configure `/proc/sys/net/ipv4/ip_local_port_range`, defaulting to `1024 65535` which seems to match what has been there since at least 5 years (!) - I couldn't trace it any further than: https://github.com/cloudfoundry/routing-release/commit/af09610f04e6430c0acdb5206b2aa2ebf226724a#diff-504a89f0b7db13eb0185c6c16ec7ee5c8822573f160db84b7fa56138baf11930R24

Anyway 1024 as a lower bound is much lower than typical defaults which seem to usually start at around 32768 (e.g.: https://tldp.org/LDP/solrhe/Securing-Optimizing-Linux-RH-Edition-v1.3/chap6sec70.html) - however the reasons for doing so seem reasonable enough.

Unless... one doesn't also exclude from that local port range any ports that our VM needs to listen on.

Today we had a Gorouter go bad at system startup time - it wasn't able to bind to the port we had specified it to listen on, because an unrelated process on the same VM had chosen to use that port number as a ephemeral point from the local port range, for a long-lived connection.

This meant that the Gorouter kept (unsuccessfully) trying to restart.

Although a default install currently listens on port 80, which is outside of the local port range (in our case we had overridden that to larger value, e.g. 4080), the default config suggests that other ports are equally likely to be affected, e.g.:

https://github.com/cloudfoundry/routing-release/blob/develop/jobs/gorouter/spec#L41-L43

or

https://github.com/cloudfoundry/routing-release/blob/develop/jobs/gorouter/spec#L330-L332

## Possible Fix

https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt describes `ip_local_reserved_ports`, pasted below.

If this were set a the same time, and in the same manner as `ip_local_port_range` (including configurable by operators), then this should prevent such a situation from occurring:

```
ip_local_port_range - 2 INTEGERS
	Defines the local port range that is used by TCP and UDP to
	choose the local port. The first number is the first, the
	second the last local port number.
	If possible, it is better these numbers have different parity
	(one even and one odd value).
	Must be greater than or equal to ip_unprivileged_port_start.
	The default values are 32768 and 60999 respectively.

ip_local_reserved_ports - list of comma separated ranges
	Specify the ports which are reserved for known third-party
	applications. These ports will not be used by automatic port
	assignments (e.g. when calling connect() or bind() with port
	number 0). Explicit port allocation behavior is unchanged.

	The format used for both input and output is a comma separated
	list of ranges (e.g. "1,2-4,10-10" for ports 1, 2, 3, 4 and
	10). Writing to the file will clear all previously reserved
	ports and update the current list with the one given in the
	input.

	Note that ip_local_port_range and ip_local_reserved_ports
	settings are independent and both are considered by the kernel
	when determining which ports are available for automatic port
	assignments.

	You can reserve ports which are not in the current
	ip_local_port_range, e.g.:

	$ cat /proc/sys/net/ipv4/ip_local_port_range
	32000	60999
	$ cat /proc/sys/net/ipv4/ip_local_reserved_ports
	8080,9148

	although this is redundant. However such a setting is useful
	if later the port range is changed to a value that will
	include the reserved ports.

	Default: Empty
```

As a workaround, we'll likely use an `osconf` workaround to set that at start-up.

It is satisfying to file this bug - we've observed over the years numerous but infrequent Gorouter failures at startup, but weren't able to pin-point a reason why.  Today we were able to see another process with `lsof` connected to something else using our port, which led us down this path of discovery. Credit to @bg-govau

**Comment by jrussett** [comment]
Hi @ae-govau,

Thanks for taking the time to write this issue.

## Related issue?
This problem seems awfully familiar, if not the same, to GH Issue #184. 

There is a doc in this repo, [Known Issue: TCP Router Fails when Port Conflicts with Local Process](https://github.com/cloudfoundry/routing-release/blob/develop/docs/tcp-router-port-conflict-known-issue.md) with a lot more information.

## Solutioning
To address the collisions, we came up with a new bosh job property `routing_api.reserved_system_component_ports`, which allows users to configure reserved ports that should not be present in the foundation's router group, thereby preventing another process from trying to claim a port in use. 

This feature has not yet been released and was only recently merged in PR #202.


I believe that once a new version of routing-release has been finalized, this should resolve your issue. 

Thanks,
@jrussett

**Comment by ghost** [comment]
Hi @jrussett , I did see #184, and while it shares a lot of the same keywords, I think it's quite a different issue.

1. This issue relates to (and can be fixed by) the prestart script for the `gorouter` component, not the `tcp-router` (although I wouldn't be surprised if it were similarly affected, I didn't check if it modifies the kernel `ip_local_port_range` in the same way). #202 doesn't change `gorouter` job behavior so far as I can tell.
2. #184 relates to conflicts with other listening ports on the same VM.  Since `tcp-router` (as I understand it, as we haven't had cause to use it) allows users to call an API to dynamically map new ports, it needs to be able to check for conflicts.  On a standard `gorouter`, all listening ports are statically declared in BOSH configuration - and that's the case we see intermittent failures on.

For example, our `gorouter` was configured to listen on port 4080. When it failed to start, we noticed the following:

```
gorouter/<uuid>:~# sudo lsof -i:4080
COMMAND   PID   USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
rsyslogd 5065 syslog   37u  IPv4  26615      0t0  TCP localhost:shell->localhost:4880 (ESTABLISHED)
blackbox 5555 syslog   43u  IPv4  26614      0t0  TCP localhost:4080->localhost:shell (ESTABLISHED)
```

Note the port conflict is caused not by another process listening on port 4080, but rather by another process (`blackbox`) randomly choosing 4080 as the ephemeral port for the local side of a TCP connection to `rsyslogd` which is running on the same host. This connection must have initiated itself *after* the `prestart` script was run (which lowered the start of `ip_local_port_range` to 1024) and before the start script for the `gorouter` was run.

As I understand it, the TCP stack is free to choose any port that is in `ip_local_port_range` as the local side of a connection, and presumably a key reason why the default is so much higher (32678) is to deconflict with normal listen ports which are typically much lower.

If the `prestart` script set `ip_local_reserved_ports` to the set of ports that we know `gorouter` needs to listen on before modifying `ip_local_port_range` this ought prevent this class of error occurring. I suspect a similar solution for `tcp-router` (which allows user modification of listen ports) would be more complex.
